crowbar_framework/app/models/api/upgrade.rb

```diff
@@ -535,6 +572,81 @@ def cancel
         upgrade_status.initialize_state
       end
 
+      def upgrade_nodes_disruptive(to_upgrade)
+        to_upgrade.each do |element|
+          # If role has a cluster assigned ugprade that cluster (reuse the non-disruptive code here)
+          if ServiceObject.is_cluster?(element)
+            cluster = ServiceObject.cluster_name element
+            cluster_env = "pacemaker-config-#{cluster}"
+            founder = ::Node.find(
+              "pacemaker_founder:true AND " \
+              "pacemaker_config_environment:#{cluster_env}"
+            ).first
+            upgrade_cluster founder, cluster_env
+          else
+            # if role has single node(s) assigned upgrade those nodes (optionally all at once)
+            # FIXME: could we do it in paralel?
+            upgrade_one_node element
+          end
+        end
+
+        # and handle those as part of the compute node upgrade
+      end
+
+      # Go through active proposals and return elements with assigned roles in the right order
+      def upgradable_elements_of_proposals(proposals)
```

Comment:
I somehow agree with Hound here that this is a bit to complex. I am currently experimenting a bit with this to see if we could make it a bit simpler.